book/src/rfcs/0002-nixos-on-asterinas.md

```diff
@@ -0,0 +1,151 @@
+
+```nix
+{ config, lib, pkgs, ... }: {
+  options = {
+    asterinas.splash = lib.mkOption {
+      type = lib.types.path;
+      default = /asterinas/splash.png;
+    };
+    asterinas.kernel = lib.mkOption {
+      type = lib.types.path;
+      default = /asterinas/kernel;
+    };
+    asterinas.initramfs = lib.mkOption {
+      type = lib.types.path;
+      default = pkgs.makeInitrd {
+        compressor = "gzip";
+        contents = [
+          {
+            object = "${pkgs.busybox}/bin";
+            symlink = "/bin";
+          }
+          {
+            object = "${config.asterinas.initramfs-init}";
+            symlink = "/init";
+          }
+        ];
+      };
+    };
+  };
+
+  config = {
+    boot.loader.grub.enable = true;
+    boot.loader.grub.efiSupport = true;
+    boot.loader.grub.device = "nodev";
+    boot.loader.grub.efiInstallAsRemovable = true;
+    boot.loader.grub.splashImage = config.asterinas.splash;
+
+    boot.initrd.enable = false;
+    boot.kernel.enable = false;
+    boot.loader.grub.extraInstallCommands = ''
+      echo "Executing more commands after grub installation.."
+    '';
+    boot.postBootCommands = ''
+      echo "Executing more commands after booting.."
+    '';
+
+    system.systemBuilderCommands = ''
+      echo "Building the kernel and initrd.."
+    '';
+    systemd.enableCgroupAccounting = false;
+
+    environment.defaultPackages = [ ];
+
+    system.nixos.distroName = "Asterinas";
+  };
+}
+```
+
+This configuration demonstrates how to modify some aspects of the NixOS installer ISO, such as the kernel, initramfs, and GRUB configuration, without needing to fork the entire NixOS build system. This approach minimizes the amount of project-specific code we need to maintain for our NixOS-based Asterinas distro.
+
+A further significant advantage of NixOS is the portability of [Nix](https://nixos.org/download/#nix-install-linux), its package manager (not Nix, the programming language), across various Linux distros and even macOS.
+
+Unlike most package managers, which are tightly coupled to their parent distros (e.g., Pacman for Arch, `dpkg` for Debian, `rpm` for Red Hat-based systems), Nix operates independently. Nix packages are installed in the `/nix/store` directory, ensuring they do not conflict with native packages. This portability is invaluable for debugging: if a package in our NixOS-based distro encounters an issue, we can replicate the exact same package environment on any Linux development machine, greatly simplifying troubleshooting.
+
+### The New Top-Level Sub-Project
+
+Given these compelling rationales, we propose the creation of a new top-level sub-project named _NixOS on Asterinas (NOSA)_. This sub-project will reside in a new `distro/` directory at the project root.
```

Comment:
This section also does not explain how files are (planned to be) organized under the new `distro/` sub-project at the implementation level.

 - ArchLinux creates one GitLab repository for each package it supported, e.g, https://gitlab.archlinux.org/archlinux/packaging/packages/gcc.
 - Gentoo creates a directory under the category's directory (e.g., `app-admin`, `dev-lang`, see https://github.com/gentoo/gentoo/), e.g., https://github.com/gentoo/gentoo/tree/master/dev-lang/rust.

Perhaps there are not only packages, but also other items, such as the script that builds the ISO.

We cannot decide everything here but we should say something?

Reply:
I want to emphasize that this RFC should provide the right level of details. Implementation details are subject to change and should be considered out of scope for an RFC, which is an immutable, historic record. For example, a Rust RFC is almost always focused on describing the proposed language feature, not how the language feature is implemented.

Again, I will slightly expand this section to give the reader a more concrete idea on how the "NixOS on Asterinas" distro may be built upon the vanilla NixOS.